File: backend/PushOverAction.py
import httplib
import logging
import urllib

from google.appengine.ext import db

from datamodels import Action


class PushOverAction(Action):
    token = db.StringProperty(required=True)
    userkeys = db.StringListProperty()
    custom_sound = db.StringProperty(default='')

    def perform_action(self):
        if not self.enabled:
            return

        self.update_performed()

        if self.custom_sound != '':
            sound = self.custom_sound
        else:
            if self.failure_action:
                sound = 'falling'
            else:
                sound = 'bike'

        for user_key in self.userkeys:
            logging.info('PushOverAction sending message')
            conn = httplib.HTTPSConnection('api.pushover.net:443')
            conn.request('POST', '/1/messages.json',
                         urllib.urlencode({
                             'token': self.token,
                             'user': user_key,
                             'message': self.message,
                             'sound': sound
                         }), {'Content-type': 'application/x-www-form-urlencoded'})
            conn.getresponse()

File: backend-go/actions/pushoverAction.go
package actions

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

type PushoverAction struct {
	ActionData
	IsFailureAction bool
	Token           string
	UserKeys        []string
	CustomSound     string
}

type pushoverApiData struct {
	Token   string `json:"token"`
	User    string `json:"user"`
	Message string `json:"message"`
	Sound   string `json:"sound"`
}

func (a *PushoverAction) Run() error {
	if !a.Enabled {
		return nil
	}

	a.UpdateLastPerformed()

	var sound string
	if a.CustomSound != "" {
		sound = a.CustomSound
	} else {
		if a.IsFailureAction {
			sound = "falling"
		} else {
			sound = "bike"
		}
	}

	client := &http.Client{
		Timeout: 10 * time.Second,
	}

	var errors []error
	for _, user := range a.UserKeys {
		data, err := json.Marshal(pushoverApiData{
			Token:   a.Token,
			User:    user,
			Message: a.Message,
			Sound:   sound})

		if err != nil {
			errors = append(errors, err)
			continue
		}

		// TODO use goroutine
		response, err := client.Post("https://api.pushover.net/1/messages.json", "application/json", bytes.NewBuffer(data))

		if err != nil {
			errors = append(errors, err)
			continue
		}

		if response.StatusCode != http.StatusOK {
			errors = append(errors, fmt.Errorf("pushover service responded with %d", response.StatusCode))
		}
	}

	if len(errors) > 0 {
		return fmt.Errorf("pushover service responded with %d errors", len(errors))
	}

	return nil
}
